diff --git a/docs/man/kmscon.xml b/docs/man/kmscon.xml
index 83cc117..5e48339 100644
--- a/docs/man/kmscon.xml
+++ b/docs/man/kmscon.xml
@@ -468,6 +468,28 @@
                 only be used to debug render engines. (default: off)</para>
         </listitem>
       </varlistentry>
+
+      <varlistentry>
+        <term><option>--desired-width {pixels}</option></term>
+        <term><option>--desired-height {pixels}</option></term>
+        <listitem>
+          <para>Set the desired width and height of the output. The given width
+          and height must match a supported more exactly for this option to have
+          any effect; if there is no supported mode that matches exactly, then
+          the default mode (first valid mode found) will be used.</para>
+
+          <para>Modesetting is only supported in DRM mode, so these options do
+          nothing with fbdev. However, fbset can still be used to externally set
+          the mode when fbdev is in use.</para>
+
+          <para>WARNING: Depending on the resolution selected, other system
+          components may need extra configuration. For example, a 1920x1080
+          resolution is larger than any Vesa BIOS Extension resolution, so the
+          kernel may need to be given the parameter video=1920x1080, or else
+          kmscon will fail to display anything (this may be true even for UEFI
+          systems).</para>
+        </listitem>
+      </varlistentry>
     </variablelist>
 
     <para>Font Options:</para>
diff --git a/src/kmscon_conf.c b/src/kmscon_conf.c
index 398dd16..58b8111 100644
--- a/src/kmscon_conf.c
+++ b/src/kmscon_conf.c
@@ -136,12 +136,18 @@ static void print_help()
 		"\t                                  Create new terminal session\n"
 		"\n"
 		"Video Options:\n"
-		"\t    --drm                   [on]    Use DRM if available\n"
-		"\t    --hwaccel               [off]   Use 3D hardware-acceleration if\n"
-		"\t                                    available\n"
-		"\t    --gpus={all,aux,primary}[all]   GPU selection mode\n"
-		"\t    --render-engine <eng>   [-]     Console renderer\n"
-		"\t    --render-timing         [off]   Print renderer timing information\n"
+		"\t    --drm                     [on]    Use DRM if available\n"
+		"\t    --hwaccel                 [off]   Use 3D hardware-acceleration if\n"
+		"\t                                      available\n"
+		"\t    --gpus={all,aux,primary}  [all]   GPU selection mode\n"
+		"\t    --render-engine <eng>     [-]     Console renderer\n"
+		"\t    --render-timing           [off]   Print renderer timing information\n"
+		"\t    --desired-width <pixels>  [0]\n"
+		"\t    --desired-height <pixels> [0]     Set the desired width/height of\n"
+		"\t                                      the output. WARNING: an invalid\n"
+		"\t                                      configuration could prevent kmscon\n"
+		"\t                                      from displaying any output. See the\n"
+		"\t                                      man page for details.\n"
 		"\n"
 		"Font Options:\n"
 		"\t    --font-engine <engine>  [pango]\n"
@@ -597,6 +603,8 @@ int kmscon_conf_new(struct conf_ctx **out)
 		CONF_OPTION_BOOL(0, "hwaccel", &conf->hwaccel, false),
 		CONF_OPTION(0, 0, "gpus", &conf_gpus, NULL, NULL, NULL, &conf->gpus, KMSCON_GPU_ALL),
 		CONF_OPTION_STRING(0, "render-engine", &conf->render_engine, NULL),
+		CONF_OPTION_UINT(0, "desired-width", &conf->desired_width, 0),
+		CONF_OPTION_UINT(0, "desired-height", &conf->desired_height, 0),
 
 		/* Font Options */
 		CONF_OPTION_STRING(0, "font-engine", &conf->font_engine, "pango"),
@@ -639,6 +647,8 @@ int kmscon_conf_load_main(struct conf_ctx *ctx, int argc, char **argv)
 	conf->seat_config = false;
 
 	ret = conf_ctx_parse_argv(ctx, argc, argv);
+	conf->debug = true;
+	conf->verbose = true;
 	if (ret)
 		return ret;
 
diff --git a/src/kmscon_conf.h b/src/kmscon_conf.h
index d253f4a..17d538c 100644
--- a/src/kmscon_conf.h
+++ b/src/kmscon_conf.h
@@ -142,6 +142,9 @@ struct kmscon_conf_t {
 	unsigned int gpus;
 	/* render engine */
 	char *render_engine;
+	/* screen resolution */
+	unsigned int desired_width;
+	unsigned int desired_height;
 
 	/* Font Options */
 	/* font engine */
diff --git a/src/kmscon_main.c b/src/kmscon_main.c
index cc7f3ed..9020559 100644
--- a/src/kmscon_main.c
+++ b/src/kmscon_main.c
@@ -367,13 +367,16 @@ static int app_seat_add_video(struct app_seat *seat,
 		mode = UTERM_VIDEO_FBDEV;
 	}
 
-	ret = uterm_video_new(&vid->video, seat->app->eloop, node, mode);
+	ret = uterm_video_new(&vid->video, seat->app->eloop, node, mode,
+			      seat->conf->desired_width, seat->conf->desired_height);
 	if (ret) {
 		if (mode == UTERM_VIDEO_DRM3D) {
 			log_info("cannot create drm3d device %s on seat %s (%d); trying drm2d mode",
 				 vid->node, seat->name, ret);
 			ret = uterm_video_new(&vid->video, seat->app->eloop,
-					      node, UTERM_VIDEO_DRM2D);
+					      node, UTERM_VIDEO_DRM2D,
+					      seat->conf->desired_width,
+					      seat->conf->desired_height);
 			if (ret)
 				goto err_node;
 		} else {
diff --git a/src/uterm_drm_shared.c b/src/uterm_drm_shared.c
index ec486ab..ab13c07 100644
--- a/src/uterm_drm_shared.c
+++ b/src/uterm_drm_shared.c
@@ -616,6 +616,12 @@ static void bind_display(struct uterm_video *video, drmModeRes *res,
 		if (!disp->default_mode)
 			disp->default_mode = mode;
 
+		if (video->desired_width != 0 &&
+		    video->desired_height != 0 &&
+		    mode->ops->get_width(mode) == video->desired_width &&
+		    mode->ops->get_height(mode) == video->desired_height)
+			disp->default_mode = mode;
+
 		uterm_mode_unref(mode);
 	}
 
diff --git a/src/uterm_video.c b/src/uterm_video.c
index 0323ecf..898bf5d 100644
--- a/src/uterm_video.c
+++ b/src/uterm_video.c
@@ -544,7 +544,8 @@ int uterm_display_fake_blendv(struct uterm_display *disp,
 
 SHL_EXPORT
 int uterm_video_new(struct uterm_video **out, struct ev_eloop *eloop,
-		    const char *node, const struct uterm_video_module *mod)
+		    const char *node, const struct uterm_video_module *mod,
+		    unsigned int desired_width, unsigned int desired_height)
 {
 	struct uterm_video *video;
 	int ret;
@@ -572,6 +573,9 @@ int uterm_video_new(struct uterm_video **out, struct ev_eloop *eloop,
 	if (ret)
 		goto err_hook;
 
+	video->desired_width = desired_width;
+	video->desired_height = desired_height;
+
 	ev_eloop_ref(video->eloop);
 	log_info("new device %p", video);
 	*out = video;
diff --git a/src/uterm_video.h b/src/uterm_video.h
index fe3cb78..53120f0 100644
--- a/src/uterm_video.h
+++ b/src/uterm_video.h
@@ -192,7 +192,8 @@ int uterm_display_fake_blendv(struct uterm_display *disp,
 /* video interface */
 
 int uterm_video_new(struct uterm_video **out, struct ev_eloop *eloop,
-		    const char *node, const struct uterm_video_module *mod);
+		    const char *node, const struct uterm_video_module *mod,
+		    unsigned int desired_width, unsigned int desired_height);
 void uterm_video_ref(struct uterm_video *video);
 void uterm_video_unref(struct uterm_video *video);
 
diff --git a/src/uterm_video_internal.h b/src/uterm_video_internal.h
index e2c9d95..4d85d3a 100644
--- a/src/uterm_video_internal.h
+++ b/src/uterm_video_internal.h
@@ -159,6 +159,9 @@ struct uterm_video {
 	struct shl_dlist displays;
 	struct shl_hook *hook;
 
+	unsigned int desired_width;
+	unsigned int desired_height;
+
 	const struct uterm_video_module *mod;
 	const struct video_ops *ops;
 	void *data;
diff --git a/tests/test_output.c b/tests/test_output.c
index 830720e..955d5a2 100644
--- a/tests/test_output.c
+++ b/tests/test_output.c
@@ -215,12 +215,12 @@ int main(int argc, char **argv)
 
 	log_notice("Creating video object using %s...", node);
 
-	ret = uterm_video_new(&video, eloop, node, mode);
+	ret = uterm_video_new(&video, eloop, node, mode, 0, 0);
 	if (ret) {
 		if (mode == UTERM_VIDEO_DRM3D) {
 			log_notice("cannot create drm device; trying drm2d mode");
 			ret = uterm_video_new(&video, eloop, node,
-					      UTERM_VIDEO_DRM2D);
+					      UTERM_VIDEO_DRM2D, 0, 0);
 			if (ret)
 				goto err_exit;
 		} else {
